File: packages/native-helpers/swift-helper/Sources/SwiftHelper/RecordingAmbianceController.swift
final class RecordingAmbianceController {
    typealias PlaySound = (String) async throws -> Void

    private let gate = AsyncMutex()
    private let playSound: PlaySound
    private let muteSystemAudio: () -> Bool
    private let restoreSystemAudio: () -> Bool
    private let log: (String) -> Void

    init(
        playSound: @escaping PlaySound,
        muteSystemAudio: @escaping () -> Bool,
        restoreSystemAudio: @escaping () -> Bool,
        log: @escaping (String) -> Void
    ) {
        self.playSound = playSound
        self.muteSystemAudio = muteSystemAudio
        self.restoreSystemAudio = restoreSystemAudio
        self.log = log
    }

    func startRecording(muteSystemAudio shouldMute: Bool, muteSounds: Bool) async -> Bool {
        await gate.withLock { [self] in
            if !muteSounds {
                await playBestEffort(named: "rec-start")
            }
            return !shouldMute || muteSystemAudio()
        }
    }

    func stopRecording(wasMuted: Bool, muteSounds: Bool) async -> Bool {
        await gate.withLock { [self] in
            let success = !wasMuted || restoreSystemAudio()
            if !muteSounds {
                await playBestEffort(named: "rec-stop")
            }
            return success
        }
    }

    private func playBestEffort(named soundName: String) async {
        do {
            try await playSound(soundName)
        } catch {
            log("[IOBridge] Error playing \(soundName): \(error.localizedDescription)")
        }
    }
}

File: packages/native-helpers/windows-helper/src/Services/RecordingAmbianceController.cs
using System;
using System.Threading;
using System.Threading.Tasks;

namespace WindowsHelper.Services
{
    internal sealed class RecordingAmbianceController
    {
        private readonly SemaphoreSlim gate = new(1, 1);
        private readonly Func<string, Task> playSound;
        private readonly Func<bool> muteSystemAudio;
        private readonly Func<bool> restoreSystemAudio;
        private readonly Action<string> log;

        internal RecordingAmbianceController(
            Func<string, Task> playSound,
            Func<bool> muteSystemAudio,
            Func<bool> restoreSystemAudio,
            Action<string> log)
        {
            this.playSound = playSound;
            this.muteSystemAudio = muteSystemAudio;
            this.restoreSystemAudio = restoreSystemAudio;
            this.log = log;
        }

        internal async Task<bool> StartRecordingAsync(bool shouldMute, bool muteSounds)
        {
            await gate.WaitAsync();
            try
            {
                if (!muteSounds)
                {
                    await PlayBestEffortAsync("rec-start");
                }

                return !shouldMute || muteSystemAudio();
            }
            finally
            {
                gate.Release();
            }
        }

        internal async Task<bool> StopRecordingAsync(bool wasMuted, bool muteSounds)
        {
            await gate.WaitAsync();
            try
            {
                var success = !wasMuted || restoreSystemAudio();
                if (!muteSounds)
                {
                    await PlayBestEffortAsync("rec-stop");
                }

                return success;
            }
            finally
            {
                gate.Release();
            }
        }

        private async Task PlayBestEffortAsync(string soundName)
        {
            try
            {
                await playSound(soundName);
            }
            catch (Exception ex)
            {
                log($"[RpcHandler] Error playing {soundName}: {ex.Message}");
            }
        }
    }
}
